spras/containers.py

```diff
@@ -148,6 +150,42 @@ def run_container(framework: str, container_suffix: str, command: List[str], vol
     else:
         raise ValueError(f'{framework} is not a recognized container framework. Choose "docker" or "singularity".')
 
+def run_container_and_log(name: str, framework: str, container_suffix: str, command: List[str], volumes: List[Tuple[PurePath, PurePath]], working_dir: str, environment: str = 'SPRAS=True'):
+    """
+    Runs a command in the container using Singularity or Docker with associated pretty printed messages.
+    @param framework: singularity or docker
+    @param container_suffix: name of the DockerHub container without the 'docker://' prefix
+    @param command: command to run in the container
+    @param volumes: a list of volumes to mount where each item is a (source, destination) tuple
+    @param working_dir: the working directory in the container
+    @param environment: environment variables to set in the container
+    @return: output from Singularity execute or Docker run
+    """
+    print('Running {} on framework "{}" with command: {}'.format(name, framework, ' '.join(command)), flush=True)
+    try:
+        out = run_container(framework=framework, container_suffix=container_suffix, command=command, volumes=volumes, working_dir=working_dir, environment=environment)
+        if out is not None:
+            if isinstance(out, list):
+                out = ''.join(out)
+            elif isinstance(out, dict):
+                if 'message' in out:
+                    # singularity message - lets print it.
+                    if 'return_code' in out and not out['return_code'] == 0:
+                        print(f"(Program exited with non-zero exit code '{out['return_code']}')")
+                    out = ''.join(out['message'])
+                else:
+                    print("Note: This is an unknown message format - if you want this pretty printed, please file out an issue.")
+                    out = str(out)
+            elif not isinstance(out, str):
+                out = str(out, "utf-8")
+            print(indent(out))
+    except docker.errors.ContainerError as err:
+        print(f"(Command formatted as list: `{err.command}`)")
+        print(f"An unexpected non-zero exit status ({err.exit_status}) inside the docker image {err.image} occurred:")
+        err = str(err.stderr if err.stderr is not None else "", "utf-8")
```

Comment:
The primary piece about this PR, as mentioned in slack, is that these `err: str` are constructed per-runtime. In another PR, we can have these errors be filtered for messages that we know are thrown by containers in certain situations, and we can handle them appropriately, while sufficiently raising exceptions of unknown errors to stop the Snakemake runtime and report those issues to SPRAS as a bug.

The current way that errors are handled have caused Aden and I a good amount of trouble when wrapping algorithms, since it's quite hard to tell that the output even is an error, and, if the algorithm that's being wrapped happens to touch the output file, it can sometimes make tests pass when they should fail.